test/not_null.t.cpp

```diff
@@ -1486,23 +1486,23 @@ CASE( "not_null<>: Able to deduce element_type of user-defined smart pointers ev
 CASE( "not_null<>: Hashes match the hashes of the wrapped pointer" )
 {
     int i = 42;
-    gsl::not_null< const int* > raw_pointer = gsl::make_not_null( &i );
-    EXPECT( std::hash< gsl::not_null< const int* > >()(raw_pointer) == std::hash< const int* >()( gsl::as_nullable( raw_pointer ) ) );
-    gsl::not_null< std::unique_ptr< int > > unique_pointer = gsl::make_not_null( std::make_unique< int >(43) );
-    EXPECT( std::hash< gsl::not_null< std::unique_ptr< int > > >()(unique_pointer) == std::hash< std::unique_ptr< int > >()( gsl::as_nullable( unique_pointer) ) );
-    gsl::not_null< std::shared_ptr< int > > shared_pointer = gsl::make_not_null( std::make_shared< int >(43) );
-    EXPECT( std::hash< gsl::not_null< std::shared_ptr< int > > >()(shared_pointer) == std::hash< std::shared_ptr< int > >()( gsl::as_nullable( shared_pointer) ) );
+    not_null< const int* > raw_pointer = make_not_null( &i );
+    EXPECT( std::hash< not_null< const int* > >()(raw_pointer) == std::hash< const int* >()( as_nullable( raw_pointer ) ) );
+    not_null< std::unique_ptr< int > > unique_pointer = make_not_null( make_unique< int >(43) );
+    EXPECT( std::hash< not_null< std::unique_ptr< int > > >()(unique_pointer) == std::hash< std::unique_ptr< int > >()( as_nullable( unique_pointer) ) );
+    not_null< std::shared_ptr< int > > shared_pointer = make_not_null( make_shared< int >(43) );
+    EXPECT( std::hash< not_null< std::shared_ptr< int > > >()(shared_pointer) == std::hash< std::shared_ptr< int > >()( as_nullable( shared_pointer) ) );
 }
 
 CASE( "not_null<>: Hash functor disabled for non-hashable pointers and enabled for hashable pointers" )
 {
-    EXPECT(( std::is_default_constructible< std::hash< gsl::not_null< int* > > >::value ));
-    EXPECT(( std::is_default_constructible< std::hash< gsl::not_null< std::unique_ptr< int > > > >::value ));
-    EXPECT(( std::is_default_constructible< std::hash< gsl::not_null< std::shared_ptr< int > > > >::value ));
+    EXPECT(( std::is_default_constructible< std::hash< not_null< int* > > >::value ));
+    EXPECT(( std::is_default_constructible< std::hash< not_null< std::unique_ptr< int > > > >::value ));
+    EXPECT(( std::is_default_constructible< std::hash< not_null< std::shared_ptr< int > > > >::value ));
 #if gsl_CPP11_140
     // doesn't even compile in VS 2013, but at least the error message is somewhat descriptive:
     //     error C2338: The C++ Standard doesn't provide a hash for this type.
-    EXPECT_NOT(( std::is_default_constructible< std::hash< gsl::not_null< NormalPtr< int > > > >::value ));
+    EXPECT_NOT(( std::is_default_constructible< std::hash< not_null< NormalPtr< int > > > >::value ));
```

Comment:
Due to the C++17 wording changes you cited, I think the `EXPECTS_NOT()` check here should be guarded with `#if gsl_STDLIB_CPP17_OR_GREATER`. As far as I can see, the AppleClang tests succeed with `-std=c++17` but fail with e.g. `-std=c++11`.

Reply:
I just pushed a test commit with `#if gsl_STDLIB_CPP14_OR_GREATER && ! gsl_BETWEEN(gsl_COMPILER_MSVC_VERSION, 1, 141)`. Let's see whether that works on all tested platforms.

(I'll squash the test commits when the PR is ready to merge.)

Reply:
(No need to squash, I'll do that upon merging.)

Reply:
So, it seems like VS 2015 is the only exception among the tested compilers. All others (including VS 2017 with `/std:c++14`) define `std::hash` as required by the C++17 specification.